spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/SparkWrite.java

```diff
@@ -312,7 +320,13 @@ public void commit(WriterCommitMessage[] messages) {
         return;
       }
 
-      ReplacePartitions dynamicOverwrite = table.newReplacePartitions();
+      ReplacePartitions dynamicOverwrite;
+      if (branch != null) {
+        dynamicOverwrite = table.newReplacePartitions().toBranch(branch);
+      } else {
+        dynamicOverwrite = table.newReplacePartitions();
+      }
```

Comment:
Same as above we can remove the if/else after the change for a branch to always be defined.